Apple: Enable Imaging support for Embedded Platforms

dgovil opened the pull request:
### Description of Change(s)

This PR ports Thor's Imaging work onto the latest USD dev branch and enables imaging for iOS and visionOS. It also includes some patches from Maddy Adams and me to suit the current state of USD and its dependencies.

There is a single patch to MaterialX to enable support with USD not using OpenGL builds.

Additionally, as requested by Nick Porcino, I have added support for shared memory texture handling on the Mac as well where applicable.

I've tested this on macOS with Apple Silicon but do not have Intel macs to test with. I've also tested building this for both iOS and visionOS.

(Jess I believe you'd tracked this as #USD-9841)


### Fixes Issue(s)
-

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [ ] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

dgovil committed:
Apple: Enable Imaging support for Embedded Platforms

This PR ports Thor's Imaging work onto the latest USD dev branch. It also includes some patches from Maddy Adams and me to suit the current state of USD and its dependencies.

There is a single patch to MaterialX to enable support with USD not using OpenGL builds.

Additionally, as requested by Nick Porcino, I have added support for shared memory texture handling on the Mac as well where applicable.

I've tested this on macOS with Apple Silicon but do not have Intel macs to test with. I've also tested building this for both iOS and visionOS.

jesschimein commented:
Filed as internal issue #USD-9955

jesschimein commented:
/AzurePipelines run

tcauchois commented on pxr/imaging/hgiMetal/blitCmds.mm:
```diff
@@ -203,7 +203,11 @@
+#if defined(ARCH_OS_OSX)
     mtlDesc.resourceOptions = MTLResourceStorageModeManaged;
+#else
+    mtlDesc.resourceOptions = MTLResourceStorageModeShared;
```
This line seems like it comes up a lot in this PR; it would be nice if we could encapsulate the #if #else in a function, on hgi.mm or capabilities.mm or something.

dgovil replied:
yeah good callout. I'll take a pass at notes next week and hit that up.

dgovil replied:
Should be addressed in the latest commit. I moved it into capabilities as preferredStorageMode, to mirror the existing defaultStorageMode.

dgovil commented:
~~I added a commit that moves all the various shader gen backends behind their respective ifdefs~~

dgovil committed:
Set preferredStorageMode in capabilities so we don't duplicate the logic in multiple places

dgovil commented:
@davidgyu actually never mind, @ld_kerley caught some issues so I backed out my shader gen changes. I think we need to revisit it on the MaterialX side and then follow up in OpenUSD later. I'll leave the MaterialX build patch in for that reason.

jesschimein commented:
/AzurePipelines run

furby-tm committed:
Fix Garch opengl check, add ARC conditionals to HgiMetal.
* Note blitCmds.mm is still using the wrong MTLResource
  type on L.361, which should be MTLBuffer, but that issue
  is addressed in the following PR:
  https://github.com/PixarAnimationStudios/OpenUSD/pull/3226

Signed-off-by: furby™ <devs@wabi.foundation>

dgovil commented:
Just a heads up that I added some changes from @furby-tm to fix a few minor issues they found.
One thing too though is that it guards the releases behind an arc check, and if we can merge in https://github.com/PixarAnimationStudios/OpenUSD/pull/3226 then we can potentially remove those as well.

jesschimein commented:
/AzurePipelines run

dgovil committed:
Resolve merge conflict

jesschimein commented:
/AzurePipelines run

dgovil committed:
Fix a compiler preprocessing issue in glPlatformContextDarwin.mm that popped up after rebasing on latest dev
